### DISCUSSION TITLE
Issues about MessageArray in python

### DISCUSSION THREAD

**Discussion by LangLEvoI** [FLAMEGPU/FLAMEGPU2#1109] [Q&A] [answered]
Hi there, based on the [Circle tutorial](https://docs.flamegpu.com/tutorial/index.html), I'm using `Agent Python` to build a simple model to better understand Flame GPU 2.

An agent could randomly pick up another agent, replicating another one's status in my new model.

But some errors occurred when I tried to use `MessageArray`. Could I fix it somewhere?

First, the definition of 1D message array AND agent functions:

```python
message = model.newMessageArray('variant')
message.setLength(5)
message.newVariableInt('var')


@pyflamegpu.agent_function
def output_message(message_in: pyflamegpu.MessageNone, message_out: pyflamegpu.MessageArray):
    message_out.setIndex(pyflamegpu.getID() - 1)
    message_out.setVariableInt("var", pyflamegpu.getVariableInt("var"))

    return pyflamegpu.ALIVE


@pyflamegpu.agent_function
def input_message(message_in: pyflamegpu.MessageArray, message_out: pyflamegpu.MessageNone):
    interlocutor = pyflamegpu.random.uniformInt(0, 4)
    msg = message_in.at(interlocutor)

    pyflamegpu.setVariableInt("var", msg.getVariableInt('var'))

    return pyflamegpu.ALIVE
```

When I ran the programme, the first error:

```
---------------------------------------------------
--- JIT compile log for input_message_program ---
---------------------------------------------------
input_message_impl.cu(7): error: class "flamegpu::MessageArray3D::In::Message" has no member "getVariableInt"

1 error detected in the compilation of "input_message_program".
 
---------------------------------------------------
Traceback (most recent call last):
  File "/app/lilei/projects/flame/simple.py", line 115, in
    cuda_model.applyConfig()
  File "/app/lilei/miniconda3/envs/rapids/lib/python3.8/site-packages/pyflamegpu/pyflamegpu.py", line 7223, in applyConfig
    return _pyflamegpu.Simulation_applyConfig(self)
pyflamegpu.pyflamegpu.FLAMEGPURuntimeException: (InvalidAgentFunc) /__w/FLAMEGPU2/FLAMEGPU2/src/flamegpu/detail/JitifyCache.cu(380): Error compiling runtime agent function (or function condition) ('input_message'): function had compilation errors (see std::cout), in JitifyCache::buildProgram().
```

Thx a lot!

**Comment by Robadob** [comment, marked as answer]
This appears to be a bug within the agent python transpilation.

Two things stand out:

* The error states `MessageArray3D` but you appear to be using `MessageArray`.
* It looking for the method `getVariableInt` which should actually be `getVariable<Int>`

It's possible array messages haven't been as thoroughly tested as other types within agent python.

I'll create an issue, and try to reproduce/fix it later in the week.

Thanks for the thorough post.

**Reply by Robadob** [reply]
Small update, I've now drafted a fix (https://github.com/FLAMEGPU/FLAMEGPU2/pull/1111). But it's not a part of the codebase I'm too familiar with, so will hopefully be able to discuss it with the original author on Friday to see whether they think it's appropriate.

**Reply by mondus** [reply]
This has now been merged but will require Python lib to be built until we push a new release version.

**Reply by Robadob** [reply]
You could copy [this](https://github.com/FLAMEGPU/FLAMEGPU2/tree/master/swig/python/codegen) directory to your project and use a standalone instance of codegen as a temporary workaround.

**Reply by LangLEvoI** [reply]
It works now, many thx :-)
